compiler/src/dmd/expressionsem.d

```diff
@@ -7825,6 +7825,19 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             if (exp.to == Type.terror)
                 return setError();
         }
+        if (auto toc = exp.to.isClassHandle())
+        {
+            auto e1c = exp.e1.type.isClassHandle();
+            if (e1c && toc.classKind == ClassKind.cpp &&
+                e1c.isBaseOf(toc, null))
+            {
+                // @@@DEPRECATED_2.115@@@
+                // When turning into error, uncomment the return statement
+                exp.deprecation("dynamic cast is not implemented for `extern(C++)` classes");
+                deprecationSupplemental(exp.loc, "use `cast(T)cast(void*)obj` instead");
```

Comment:
ag0aep6g writes:

`cast(D*) &c` is the unambiguous type paint. `cast(D) cast(void*) c` can be intercepted by `opCast`.